### PR TITLE
Provide structured information in binlog for the compiler server

### PR DESCRIPTION
**Pull request by jaredpar**
This PR has three commits. The first two are refactoring commits that make it much clearer how error messages are handled in the compiler server. The last commit is the one that has functional changes that provide structured information about the server behavior into the BinLog. 

This last adds structured messages to the binlog about the behavior of the compiler server. These messages will help users, our own developers and automation understand the behavior of the compiler server within a build.

Related to https://github.com/KirillOsenkov/MSBuildStructuredLog/issues/451

### CODE REVIEW

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler, @sharwell  PTAL

File: src/Compilers/Core/MSBuildTask/ErrorString.resx
Comment (jaredpar) [inline src/Compilers/Core/MSBuildTask/ErrorString.resx:202]: I did not mean to change this file, will revert.

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (cston) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:563]: >{ [](start = 12, length = 1)

`logger.Dispose();`

**Review by cston — APPROVED** [review]

**Comment by jaredpar** [comment]
@dotnet/roslyn-compiler can I get a second review here? 

@sharwell is this going to work for your efforts?

**Comment by jaredpar** [comment]
image

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (sharwell) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:704]: ❔ Intentional?

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (sharwell) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:512]:
```suggestion
                    LogCompilationMessage(logger, CompilationKind.Tool, $"using command line tool by design '{pathToTool}'");
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (sharwell) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:523]:
```suggestion
                    LogCompilationMessage(logger, CompilationKind.Tool, $"using command line tool because we could not find client directory '{PathToManagedTool}'");
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (sharwell) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:637]:
```suggestion
                LogCompilationMessage(logger, CompilationKind.ToolFallback, "could not launch server");
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (sharwell) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:655]:
```suggestion
                    LogCompilationMessage(logger, CompilationKind.FatalError, "server reports different protocol version than build task");
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (sharwell) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:659]:
```suggestion
                    LogCompilationMessage(logger, CompilationKind.FatalError, "server reports different hash version than build task");
```

File: src/Compilers/Server/VBCSCompiler/AnalyzerConsistencyChecker.cs
Comment (sharwell) [inline src/Compilers/Server/VBCSCompiler/AnalyzerConsistencyChecker.cs:102]:
```suggestion
                    var message = $"analyzer assembly '{resolvedPath}' has MVID '{resolvedPathMvid}' but loaded assembly '{loadedAssembly.FullName}' has MVID '{loadedAssemblyMvid}'";
```

**Review by sharwell — APPROVED** [review]
Messages seem usable. Might be good to have a high-level document listing the messages so future tool authors can refer to it, similar to [**Report Analyzer Format.md**](https://github.com/dotnet/roslyn/blob/121e4b43bcaac109a960f859f42d972f7d35b803/docs/analyzers/Report%20Analyzer%20Format.md).

File: src/Compilers/Core/CommandLine/BuildProtocol.cs
Comment (RikkiGibson) [inline src/Compilers/Core/CommandLine/BuildProtocol.cs:466]: Should this comment be updated?

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (RikkiGibson) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:31]:
```suggestion
            /// the customer opted out of the compiler server
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (RikkiGibson) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:42]:
```suggestion
            /// Compilation occurred in the compiler server process
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (RikkiGibson) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:30]:
```suggestion
            /// Commpilation occurred using the command line tool by normal processes, typically because 
```

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (RikkiGibson) [inline src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:564]: Can this be null here?

File: src/Compilers/Server/VBCSCompiler/AnalyzerConsistencyChecker.cs
Comment (RikkiGibson) [inline src/Compilers/Server/VBCSCompiler/AnalyzerConsistencyChecker.cs:105]: nit: this is a good place to use `??=`

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/Core/CommandLine/BuildProtocol.cs
Comment (jaredpar) [reply src/Compilers/Core/CommandLine/BuildProtocol.cs:466]: Probably a good idea.

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (jaredpar) [reply src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:31]: I contend this isn't necessarily a typo but instead a regional-ism

File: src/Compilers/Core/MSBuildTask/ManagedCompiler.cs
Comment (jaredpar) [reply src/Compilers/Core/MSBuildTask/ManagedCompiler.cs:564]: Yes, it's not the first line of the `try` hence always possible we get here via some strange exception before it's initialized.

File: src/Compilers/Server/VBCSCompiler/AnalyzerConsistencyChecker.cs
Comment (jaredpar) [reply src/Compilers/Server/VBCSCompiler/AnalyzerConsistencyChecker.cs:105]: True

**Pull request merged**
